torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,273 @@
+#include <torch/csrc/jit/passes/memory_planning.h>
+
+#include <jit/tensorexpr/kernel.h>
+#include <torch/csrc/jit/ir/alias_analysis.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/runtime/static/ops.h>
+
+namespace torch {
+namespace jit {
+
+c10::optional<uint64_t> computeStorageSize(const Value& value) {
+  auto ttp = value.type()->cast<TensorType>();
+  if (!ttp) {
+    TORCH_WARN("out isn't a tensortype ", *value.type());
+    return c10::nullopt;
+  }
+  if (!ttp->scalarType().has_value()) {
+    TORCH_WARN(
+        "This output was profiled but didn't have a scalar type: ",
+        *ttp,
+        ", ",
+        value.debugName());
+    return c10::nullopt;
+  }
+  if (!ttp->sizes().concrete_sizes().has_value()) {
+    TORCH_WARN(
+        "This output was profiled but doesn't have sizes: ",
+        *ttp,
+        ", ",
+        value.debugName());
+    return c10::nullopt;
+  }
+
+  auto scalar_type = ttp->scalarType();
+  if (!scalar_type.has_value()) {
+    TORCH_WARN(
+        "This value doesn't have a scalar type", *ttp, ", ", value.debugName());
+    return c10::nullopt;
+  }
+
+  auto element_size = c10::elementSize(scalar_type.value());
+  // TODO: when does this fail? answer: in place mutation
+  auto numel = ttp->numel();
+  if (!numel.has_value()) {
+    TORCH_WARN("doesn't have numel", *ttp, ", ", value.debugName());
+    return c10::nullopt;
+  }
+
+  return numel.value() * element_size;
+}
+
+std::pair<std::vector<int64_t>, std::vector<int64_t>> getSizesStrides(
+    const c10::TensorTypePtr& ttp) {
+  std::vector<int64_t> sizes;
+  auto _sizes = ttp->sizes().concrete_sizes();
+  // TODO: why does this break? answer: in place mutation
+  // also %9 : Long(requires_grad=0, device=cpu) = prim::Constant[value={0}]()
+  if (_sizes.has_value() && _sizes.value().size() > 0 &&
+      _sizes.value()[0] != 0) {
+    sizes = _sizes.value();
+  } else {
+    sizes = std::vector<int64_t>{0};
+  }
+  std::vector<int64_t> strides;
+  auto _strides = ttp->strides().concrete_sizes();
+  if (_strides.has_value() && _strides.value().size() > 0 &&
+      _strides.value()[0] != 0) {
+    strides = _strides.value();
+  } else {
+    strides = at::detail::defaultStrides(sizes);
+  }
+  return std::make_pair(sizes, strides);
+}
+
+Node* insertAllocStorageNode(
+    std::shared_ptr<Graph>& graph,
+    uint64_t total_size) {
+  auto* storage = graph->create(prim::AllocateStorage, 1);
+  storage->i_(attr::total_size, total_size);
+
+  auto deviceType = jit::tensorexpr::pickDeviceType(graph);
+  if (deviceType.has_value()) {
+    storage->i_(attr::device, static_cast<int8_t>(deviceType.value().type()));
+  } else {
+    storage->i_(attr::device, static_cast<int8_t>(at::kCPU));
+  }
+  storage->insertBefore(graph->nodes().front());
+  return storage;
+}
+
+void insertAllocTensorNodes(
+    std::shared_ptr<Graph>& graph,
+    Node* storage,
+    std::unordered_map<LiveRange, Region, live_range_hash> allocations,
```

Comment:
What does LiveRange represent? I saw it it's size_t start, end, but that is not very descriptive. Could you add some comments to its definition as part of this PR?